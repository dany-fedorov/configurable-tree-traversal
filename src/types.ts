export interface TreeTypeParameters<VD = unknown, VH = unknown> {
  VertexData: VD;
  VertexHint: VH;
}

export interface Vertex<TTP extends TreeTypeParameters> {
  readonly $d: TTP['VertexData'];
  readonly $c: TTP['VertexHint'][];
}

export type MakeVertexOptions<TTP extends TreeTypeParameters> =
  TreeResolution<TTP> & Omit<IVertexContext<TTP>, 'hint'>;

export interface TraversableTree<TTP extends TreeTypeParameters> {
  makeRoot(): Vertex<TTP> | null;

  makeVertex(
    vertexHint: TTP['VertexHint'],
    options: MakeVertexOptions<TTP>,
  ): Vertex<TTP> | null;
}

export type VertexContextMap<TTP extends TreeTypeParameters> = Map<
  Vertex<TTP>,
  IVertexContext<TTP> | null
>;

export type ResolvedTreeMap<TTP extends TreeTypeParameters> = Map<
  Vertex<TTP>,
  Array<Vertex<TTP>>
>;

export interface TreeResolution<TTP extends TreeTypeParameters> {
  vertexContextMap: VertexContextMap<TTP>;
  resolvedTreeMap: ResolvedTreeMap<TTP>;
}

export interface TraversalVisitorOptions<TTP extends TreeTypeParameters>
  extends TreeResolution<TTP> {
  visitIndex: number;
  previousVisitedVertex: Vertex<TTP> | null;
  isLeafVertex: boolean;
  isRootVertex: boolean;
}

export enum TraversalVisitorCommand {
  HALT_TRAVERSAL = 'HALT_TRAVERSAL',
}

export interface TraversalVisitorResult {
  command?: TraversalVisitorCommand;
}

export type TraversalVisitor<TTP extends TreeTypeParameters> = (
  vertex: Vertex<TTP>,
  options: TraversalVisitorOptions<TTP>,
) => TraversalVisitorResult | undefined;

export interface IVertexContext<TTP extends TreeTypeParameters> {
  parentVertex: Vertex<TTP>;
  vertexHintOriginalOrderIndex: number;
  vertexHintTraversalOrderIndex: number;
  hint: TTP['VertexHint'];
}
