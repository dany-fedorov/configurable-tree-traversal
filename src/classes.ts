import type { Vertex, TreeTypeParameters } from './types';

interface CVertexProps<TTP extends TreeTypeParameters> {
  data: TTP['VertexData'];
  childrenHints: TTP['VertexHint'][];
}

export class CVertex<TTP extends TreeTypeParameters> implements Vertex<TTP> {
  readonly $d: TTP['VertexData'];
  readonly $c: TTP['VertexHint'][];

  constructor(vertexProps: CVertexProps<TTP>) {
    this.$d = vertexProps.data;
    this.$c = vertexProps.childrenHints;
  }

  static fromPlain<TTP extends TreeTypeParameters>(
    vertexPlain: Vertex<TTP>,
  ): CVertex<TTP> {
    return new CVertex<TTP>({
      data: vertexPlain.$d,
      childrenHints: vertexPlain.$c,
    });
  }

  static makePlain<TTP extends TreeTypeParameters>(
    vertexProps: CVertexProps<TTP>,
  ): Vertex<TTP> {
    return {
      $d: vertexProps.data,
      $c: vertexProps.childrenHints,
    };
  }

  static getData<TTP extends TreeTypeParameters>(
    vertex: Vertex<TTP>,
  ): TTP['VertexData'] {
    return vertex.$d;
  }

  static getChildrenHints<TTP extends TreeTypeParameters>(
    vertex: Vertex<TTP>,
  ): TTP['VertexHint'][] {
    return vertex.$c;
  }

  getChildrenHints() {
    return CVertex.getChildrenHints<TTP>(this);
  }

  getData() {
    return CVertex.getData<TTP>(this);
  }

  toPlain(): Vertex<TTP> {
    return {
      $d: this.$d,
      $c: this.$c,
    };
  }
}
