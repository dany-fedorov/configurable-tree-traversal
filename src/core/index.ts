export * from './CTTAbstractParent';
export * from './CTTRef';
export * from './MakeMutationCommandFunctionFactory';
export * from './ResolvedTree';
export * from './ResolvedTreeTypeParameters';
export * from './TraversableTree';
export * from './TraversableTreeParametersFromTraversableTree';
export * from './TraversalVisitor';
export * from './TreeTypeParameters';
export * from './Vertex';
