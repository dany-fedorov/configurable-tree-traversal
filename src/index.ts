export * from './types';
export * from './CVertex';
export * from './traversals/traverse-depth-first';
