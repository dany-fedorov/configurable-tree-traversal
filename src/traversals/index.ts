export * from './depth-first-traversal/DepthFirstTraversal';
