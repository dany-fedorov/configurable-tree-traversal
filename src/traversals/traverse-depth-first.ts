import type {
  TraversableTree,
  Vertex,
  IVertexContext,
  TraversalVisitor,
  TreeTypeParameters,
  ResolvedTreeMap,
  VertexContextMap,
  TreeResolution,
} from '../types';
import { CVertex } from '../CVertex';
import { TraversalVisitorCommand } from '../types';

export enum ChildrenOrder {
  DEFAULT = 'DEFAULT',
  REVERSED = 'REVERSED',
}

export interface DepthFirstTraversalConfig {
  childrenOrder: ChildrenOrder;
}

export const DEFAULT_DEPTH_FIRST_TRAVERSAL_CONFIG: DepthFirstTraversalConfig = {
  childrenOrder: ChildrenOrder.DEFAULT,
};

export interface TraversalResult<TTP extends TreeTypeParameters>
  extends TreeResolution<TTP> {
  rootVertex: Vertex<TTP> | null;
}

export interface DepthFirstVisitors<TTP extends TreeTypeParameters> {
  preOrderVisitor?: TraversalVisitor<TTP>;
  postOrderVisitor?: TraversalVisitor<TTP>;
  inOrderVisitor?: TraversalVisitor<TTP>;
}

export function traverseDepthFirst<
  TTP extends TreeTypeParameters = TreeTypeParameters,
>(
  tree: TraversableTree<TTP>,
  visitors: DepthFirstVisitors<TTP>,
  config: Partial<DepthFirstTraversalConfig> = DEFAULT_DEPTH_FIRST_TRAVERSAL_CONFIG,
): TraversalResult<TTP> {
  type ThisIVertexContext = IVertexContext<TTP>;
  type ThisIVertex = Vertex<TTP>;
  type ThisTraversalOrderContext = {
    visitIndex: number;
    previousVisitedVertex: ThisIVertex | null;
  };
  type VisitorsContext = {
    [K in keyof Required<DepthFirstVisitors<TTP>>]: ThisTraversalOrderContext;
  };

  const effectiveConfig =
    config === DEFAULT_DEPTH_FIRST_TRAVERSAL_CONFIG
      ? config
      : { ...DEFAULT_DEPTH_FIRST_TRAVERSAL_CONFIG, ...config };
  const vertexStack: Array<ThisIVertexContext> = [];
  const visitorsContext: VisitorsContext = {
    preOrderVisitor: {
      visitIndex: 0,
      previousVisitedVertex: null,
    },
    postOrderVisitor: {
      visitIndex: 0,
      previousVisitedVertex: null,
    },
    inOrderVisitor: {
      visitIndex: 0,
      previousVisitedVertex: null,
    },
  };
  const vertexContextMap: VertexContextMap<TTP> = new Map();
  const resolvedTreeMap: ResolvedTreeMap<TTP> = new Map();
  const postOrderNotVisitedChildrenCountMap = new Map<ThisIVertex, number>();
  const inOrderNotVisitedChildrenCountMap = new Map<ThisIVertex, number>();
  let haltTraversalFlag = false;
  const onPostOrder = !visitors?.postOrderVisitor
    ? null
    : onPostOrderProcessing;
  const onInOrder = !visitors?.inOrderVisitor ? null : onInOrderProcessing;

  const rootVertex = tree.makeRoot();
  if (rootVertex === null) {
    return { rootVertex, vertexContextMap, resolvedTreeMap };
  }

  onPreOrder(rootVertex, null);
  pushHints(rootVertex);

  while (vertexStack.length > 0 && !haltTraversalFlag) {
    const vertexContext = vertexStack.pop() as ThisIVertexContext;
    const vertex = tree.makeVertex(vertexContext.hint, {
      ...vertexContext,
      resolvedTreeMap,
      vertexContextMap,
    });
    if (vertex === null) {
      continue;
    }
    onPreOrder(vertex, vertexContext);
    pushHints(vertex);
    if (isLeafVertex(vertex)) {
      onPostOrder?.(vertex, vertexContext);
      onInOrder?.(vertex, vertexContext);
    }
  }

  function onPreOrder(
    vertex: ThisIVertex,
    vertexContext: ThisIVertexContext | null,
  ): void {
    if (vertexContext !== null) {
      let parentVertexChildren = resolvedTreeMap.get(
        vertexContext.parentVertex,
      );
      if (parentVertexChildren === undefined) {
        parentVertexChildren = [];
        resolvedTreeMap.set(vertexContext.parentVertex, parentVertexChildren);
      }
      parentVertexChildren.push(vertex);
    }
    vertexContextMap.set(vertex, vertexContext);
    visitVertex('preOrderVisitor', vertex);
  }

  function onInOrderProcessing(
    vertex: ThisIVertex,
    vertexContext: ThisIVertexContext | null,
  ): void {
    visitVertex('inOrderVisitor', vertex);
    if (!isLeafVertex(vertex) || !vertexContext?.parentVertex) {
      return;
    }
    // let curVertex = vertex;
    let curVertexContext: ThisIVertexContext | null = vertexContext;
    while (curVertexContext !== null) {
      const inOrderNotVisitedChildrenCount =
        inOrderNotVisitedChildrenCountMap.get(curVertexContext.parentVertex);
      if (inOrderNotVisitedChildrenCount === undefined) {
        return;
      }
      const newCount = inOrderNotVisitedChildrenCount - 1;
      if (newCount !== 0) {
        inOrderNotVisitedChildrenCountMap.set(
          curVertexContext.parentVertex,
          newCount,
        );
        const parentVertexContext = vertexContextMap.get(
          curVertexContext.parentVertex,
        );
        if (parentVertexContext !== undefined) {
          onInOrder?.(curVertexContext.parentVertex, parentVertexContext);
        }
        break;
      } else {
        const parentVertexContext = vertexContextMap.get(
          curVertexContext.parentVertex,
        );
        if (parentVertexContext === undefined) {
          break;
        }
        // curVertex = curVertexContext.parentVertex;
        curVertexContext = parentVertexContext;
      }
    }
  }

  function onPostOrderProcessing(
    vertex: ThisIVertex,
    vertexContext: ThisIVertexContext | null,
  ): void {
    let curVertex = vertex;
    let curVertexContext = vertexContext;
    while (curVertex !== null) {
      visitVertex('postOrderVisitor', curVertex);
      if (!curVertexContext?.parentVertex) {
        break;
      }
      const postOrderNotVisitedChildrenCount =
        postOrderNotVisitedChildrenCountMap.get(curVertexContext.parentVertex);
      if (postOrderNotVisitedChildrenCount === undefined) {
        break;
      }
      const newCount = postOrderNotVisitedChildrenCount - 1;
      if (newCount !== 0) {
        postOrderNotVisitedChildrenCountMap.set(
          curVertexContext.parentVertex,
          newCount,
        );
        break;
      } else {
        postOrderNotVisitedChildrenCountMap.delete(
          curVertexContext.parentVertex,
        );
        const parentVertexContext = vertexContextMap.get(
          curVertexContext.parentVertex,
        );
        if (parentVertexContext === undefined) {
          break;
        }
        curVertex = curVertexContext.parentVertex;
        curVertexContext = parentVertexContext;
      }
    }
  }

  function visitVertex(
    visitorOrderKey: keyof Required<DepthFirstVisitors<TTP>>,
    vertex: Vertex<TTP>,
  ): void {
    const visitorResult = visitors?.[visitorOrderKey]?.(vertex, {
      vertexContextMap,
      resolvedTreeMap,
      visitIndex: visitorsContext[visitorOrderKey].visitIndex++,
      previousVisitedVertex:
        visitorsContext[visitorOrderKey].previousVisitedVertex,
      isLeafVertex: isLeafVertex(vertex),
      isRootVertex: isRootVertex(vertex),
    });
    visitorsContext[visitorOrderKey].previousVisitedVertex = vertex;
    if (visitorResult?.command === TraversalVisitorCommand.HALT_TRAVERSAL) {
      haltTraversalFlag = true;
    }
  }

  function isRootVertex(vertex: ThisIVertex) {
    return vertex === rootVertex;
  }

  function isLeafVertex(vertex: ThisIVertex) {
    return getChildrenHints(vertex).length === 0;
  }

  function getChildrenHints(vertex: ThisIVertex) {
    return CVertex.prototype.getChildrenHints.call(vertex);
  }

  function pushHints(parentVertex: ThisIVertex): Array<ThisIVertexContext> {
    const childrenHints = getChildrenHints(parentVertex);
    const newEntries: ThisIVertexContext[] = [];
    for (let i = 0; i < childrenHints.length; i++) {
      const hintIndex =
        effectiveConfig.childrenOrder === ChildrenOrder.REVERSED
          ? i
          : childrenHints.length - i - 1;
      const hint = childrenHints[hintIndex];
      newEntries.push({
        hint,
        parentVertex,
        vertexHintOriginalOrderIndex: hintIndex,
        vertexHintTraversalOrderIndex: childrenHints.length - i - 1,
      });
    }
    vertexStack.push(...newEntries);
    if (visitors?.postOrderVisitor) {
      postOrderNotVisitedChildrenCountMap.set(parentVertex, newEntries.length);
    }
    if (visitors?.inOrderVisitor) {
      inOrderNotVisitedChildrenCountMap.set(parentVertex, newEntries.length);
    }
    return newEntries;
  }

  return { resolvedTreeMap, vertexContextMap, rootVertex };
}
