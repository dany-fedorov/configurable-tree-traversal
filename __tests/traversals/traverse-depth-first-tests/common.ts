import { Vertex } from '../../../src/Vertex';
import type {
  TraversableTree,
  TreeTypeParameters,
  VertexContent,
  TraversalVisitor,
  TraversalVisitorOptions,
} from '../../../src/types';
import {
  DEFAULT_DEPTH_FIRST_TRAVERSAL_CONFIG,
  DepthFirstTraversalConfig,
  DepthFirstVisitors,
  traverseDepthFirst,
} from '../../../src/traversals/traverse-depth-first';

export type TTP1 = TreeTypeParameters<string, string>;

export const tree1: TraversableTree<TTP1> = {
  makeRoot() {
    return Vertex.makeContent({
      data: '1',
      childrenHints: ['1', '2'],
    });
  },
  makeVertex(hint, { parentVertex }) {
    return Vertex.makeContent({
      data: [Vertex.getDataFromContent(parentVertex), hint].join('.'),
      childrenHints: Vertex.getDataFromContent(parentVertex).length > 3 ? [] : ['1', '2'],
    });
  },
};

export type Tree2TypeParameters = TreeTypeParameters<
  string | null,
  string | null
>;

const tree2Nodes: Record<string, (string | null)[]> = {
  F: ['B', 'G'],
  B: ['A', 'D'],
  D: ['C', 'E'],
  G: [null, 'I'],
  I: ['H', null],
};

const mkNodeForTree2 = (h: string | null) =>
  Vertex.makeContent<Tree2TypeParameters>({
    data: h,
    childrenHints: h === null ? [] : tree2Nodes[h] || [],
  });

/**
 * https://en.wikipedia.org/wiki/Tree_traversal#/media/File:Sorted_binary_tree_ALL_RGB.svg
 */
export const tree2: TraversableTree<Tree2TypeParameters> = {
  makeRoot() {
    return mkNodeForTree2('F');
  },
  makeVertex(hint /*{ parentVertex }*/) {
    return mkNodeForTree2(hint);
  },
};

export function testDepthFirstTree<
  TTP extends TreeTypeParameters,
>(
  tree: TraversableTree<TTP>,
  visitorKey: keyof DepthFirstVisitors<TTP>,
  config: Partial<DepthFirstTraversalConfig> = DEFAULT_DEPTH_FIRST_TRAVERSAL_CONFIG,
  visitor?: TraversalVisitor<TTP>,
) {
  const visited: Array<{
    vertex: VertexContent<TTP>;
    options: Omit<
      TraversalVisitorOptions<TTP>,
      'resolvedTreeMap' | 'vertexContextMap'
    >;
  }> = [];
  const { rootVertex, resolvedTreeMap, vertexContextMap } =
    traverseDepthFirst<TTP>(
      tree,
      {
        [visitorKey]: (
          vertex: VertexContent<TTP>,
          options: TraversalVisitorOptions<TTP>,
        ) => {
          visited.push({
            vertex,
            options: {
              visitIndex: options.visitIndex,
              isLeafVertex: options.isLeafVertex,
              isRootVertex: options.isRootVertex,
              previousVisitedVertex: options.previousVisitedVertex,
            },
          });
          return visitor?.(vertex, options);
        },
      },
      config,
    );
  return {
    visited,
    visitedData: visited.map((v) => v.vertex.$d),
    rootVertex,
    resolvedTreeMap,
    vertexContextMap,
  };
}
