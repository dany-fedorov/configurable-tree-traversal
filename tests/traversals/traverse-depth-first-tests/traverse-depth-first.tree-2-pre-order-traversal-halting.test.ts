import { testDepthFirstTree, tree2 } from './common';
import { ChildrenOrder } from '../../../src/traversals/traverse-depth-first';
import { TraversalVisitorCommand } from '../../../src/types';

test('Tree 1: Pre-order - halting', () => {
  const {
    visited,
    visitedData,
    rootVertex,
    resolvedTreeMap,
    vertexContextMap,
  } = testDepthFirstTree(
    tree2,
    'preOrderVisitor',
    {
      childrenOrder: ChildrenOrder.DEFAULT,
    },
    (vertex) => {
      if (vertex.$d === 'C') {
        return {
          command: TraversalVisitorCommand.HALT_TRAVERSAL,
        };
      }
      return;
    },
  );
  expect(visitedData).toMatchSnapshot();
  expect(visited).toMatchSnapshot();
  expect({ rootVertex, resolvedTreeMap, vertexContextMap }).toMatchSnapshot();
});
