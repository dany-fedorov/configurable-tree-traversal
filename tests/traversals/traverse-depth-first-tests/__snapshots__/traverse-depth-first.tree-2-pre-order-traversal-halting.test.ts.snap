// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree 1: Pre-order - halting 1`] = `
Array [
  "F",
  "B",
  "A",
  "D",
  "C",
]
`;

exports[`Tree 1: Pre-order - halting 2`] = `
Array [
  Object {
    "options": Object {
      "isLeafVertex": false,
      "isRootVertex": true,
      "previousVisitedVertex": null,
      "visitIndex": 0,
    },
    "vertex": Object {
      "$c": Array [
        "B",
        "G",
      ],
      "$d": "F",
    },
  },
  Object {
    "options": Object {
      "isLeafVertex": false,
      "isRootVertex": false,
      "previousVisitedVertex": Object {
        "$c": Array [
          "B",
          "G",
        ],
        "$d": "F",
      },
      "visitIndex": 1,
    },
    "vertex": Object {
      "$c": Array [
        "A",
        "D",
      ],
      "$d": "B",
    },
  },
  Object {
    "options": Object {
      "isLeafVertex": true,
      "isRootVertex": false,
      "previousVisitedVertex": Object {
        "$c": Array [
          "A",
          "D",
        ],
        "$d": "B",
      },
      "visitIndex": 2,
    },
    "vertex": Object {
      "$c": Array [],
      "$d": "A",
    },
  },
  Object {
    "options": Object {
      "isLeafVertex": false,
      "isRootVertex": false,
      "previousVisitedVertex": Object {
        "$c": Array [],
        "$d": "A",
      },
      "visitIndex": 3,
    },
    "vertex": Object {
      "$c": Array [
        "C",
        "E",
      ],
      "$d": "D",
    },
  },
  Object {
    "options": Object {
      "isLeafVertex": true,
      "isRootVertex": false,
      "previousVisitedVertex": Object {
        "$c": Array [
          "C",
          "E",
        ],
        "$d": "D",
      },
      "visitIndex": 4,
    },
    "vertex": Object {
      "$c": Array [],
      "$d": "C",
    },
  },
]
`;

exports[`Tree 1: Pre-order - halting 3`] = `
Object {
  "resolvedTreeMap": Map {
    Object {
      "$c": Array [
        "B",
        "G",
      ],
      "$d": "F",
    } => Array [
      Object {
        "$c": Array [
          "A",
          "D",
        ],
        "$d": "B",
      },
    ],
    Object {
      "$c": Array [
        "A",
        "D",
      ],
      "$d": "B",
    } => Array [
      Object {
        "$c": Array [],
        "$d": "A",
      },
      Object {
        "$c": Array [
          "C",
          "E",
        ],
        "$d": "D",
      },
    ],
    Object {
      "$c": Array [
        "C",
        "E",
      ],
      "$d": "D",
    } => Array [
      Object {
        "$c": Array [],
        "$d": "C",
      },
    ],
  },
  "rootVertex": Object {
    "$c": Array [
      "B",
      "G",
    ],
    "$d": "F",
  },
  "vertexContextMap": Map {
    Object {
      "$c": Array [
        "B",
        "G",
      ],
      "$d": "F",
    } => null,
    Object {
      "$c": Array [
        "A",
        "D",
      ],
      "$d": "B",
    } => Object {
      "hint": "B",
      "parentVertex": Object {
        "$c": Array [
          "B",
          "G",
        ],
        "$d": "F",
      },
      "vertexHintOriginalOrderIndex": 0,
      "vertexHintTraversalOrderIndex": 0,
    },
    Object {
      "$c": Array [],
      "$d": "A",
    } => Object {
      "hint": "A",
      "parentVertex": Object {
        "$c": Array [
          "A",
          "D",
        ],
        "$d": "B",
      },
      "vertexHintOriginalOrderIndex": 0,
      "vertexHintTraversalOrderIndex": 0,
    },
    Object {
      "$c": Array [
        "C",
        "E",
      ],
      "$d": "D",
    } => Object {
      "hint": "D",
      "parentVertex": Object {
        "$c": Array [
          "A",
          "D",
        ],
        "$d": "B",
      },
      "vertexHintOriginalOrderIndex": 1,
      "vertexHintTraversalOrderIndex": 1,
    },
    Object {
      "$c": Array [],
      "$d": "C",
    } => Object {
      "hint": "C",
      "parentVertex": Object {
        "$c": Array [
          "C",
          "E",
        ],
        "$d": "D",
      },
      "vertexHintOriginalOrderIndex": 0,
      "vertexHintTraversalOrderIndex": 0,
    },
  },
}
`;
